libr/anal/p/anal_ppc_cs.c

```diff
@@ -160,44 +160,48 @@ static const char* getspr(struct Getarg *gop, int n) {
 	return cspr;
 }
 
-//TODO PJ
 static void opex(RStrBuf *buf, csh handle, cs_insn *insn) {
 	int i;
-	r_strbuf_init (buf);
-	r_strbuf_append (buf, "{");
+	PJ *pj = pj_new ();
+	if (!pj) {
+		return;
+	}
+	pj_o (pj);
+	pj_ka (pj, "operands");
 	cs_sysz *x = &insn->detail->sysz;
-	r_strbuf_append (buf, "\"operands\":[");
 	for (i = 0; i < x->op_count; i++) {
-		cs_sysz_op *op = &x->operands[i];
-		if (i > 0) {
-			r_strbuf_append (buf, ",");
-		}
-		r_strbuf_append (buf, "{");
+		cs_sysz_op *op = x->operands + i;
```

Comment:
Wait, why is this sysz?

Reply:
Probably meant to be cs_ppc_op. pancake takes the git blame for this one, it likely happened when copying over opex functions, as they share very similar code between multiple plugins. Should probably open a new PR for changing that. May want to pull the opex() funcs out of the plugins themselves, too.

Reply:
> Probably meant to be cs_ppc_op. pancake takes the git blame for this one, it likely happened when copying over opex functions, as they share very similar code between multiple plugins. Should probably open a new PR for changing that. May want to pull the opex() funcs out of the plugins themselves, too.

Better fix here as it's a bug :-)
As C lacks reflection in Java, it seems impossible to have a unified opex function because we cannot dynamically know what fields a struct has.

Reply:
Could use a solution similar to CPython's PyObject with a generic struct header. So, there could exist an opcode struct header containing the type, then opex() could check the universal type member, then cast based on it, likely calling other functions based on the type.

For the record, this is well-defined behavior, not a hack. From section 3.5.2.1: Structure and union specifiers of the [C89 draft standard](http://port70.net/~nsz/c/c89/c89-draft.html): "A pointer to a structure object, suitably cast, points to its initial member (or if that member is a bit-field, then to the unit in which it resides), and vice versa. There may therefore be unnamed holes within a structure object, but not at its beginning, as necessary to achieve the appropriate alignment."

You can see how the CPython developers made their code compatible with standard C [here](https://www.python.org/dev/peps/pep-3123/).

While this may or may not be a desirable addition, it is certainly possible to add this functionality.

Reply:
> Could use a solution similar to CPython's PyObject with a generic struct header. So, there could exist an opcode struct header containing the type, then opex() could check the universal type member, then cast based on it, likely calling other functions based on the type.
> 
> For the record, this is well-defined behavior, not a hack. From section 3.5.2.1: Structure and union specifiers of the [C89 draft standard](http://port70.net/~nsz/c/c89/c89-draft.html): "A pointer to a structure object, suitably cast, points to its initial member (or if that member is a bit-field, then to the unit in which it resides), and vice versa. There may therefore be unnamed holes within a structure object, but not at its beginning, as necessary to achieve the appropriate alignment."
> 
> You can see how the CPython developers made their code compatible with standard C [here](https://www.python.org/dev/peps/pep-3123/) and the PyObject header file is [here](https://github.com/python/cpython/blob/master/Include/object.h).
> 
> While this may or may not be a desirable addition, it is certainly possible to add this functionality.

I know this way to implement polymorphism in C (C++ just does so), however we still need to **define** different functions to deal with different structs, which is exactly current situation.

Reply:
I may have misunderstood you, I just think that having this output processing in the plugins themselves is not desirable.